compiler/rustc_typeck/src/check/wfcheck.rs

```diff
@@ -201,6 +201,40 @@ pub fn check_trait_item(tcx: TyCtxt<'_>, def_id: LocalDefId) {
     };
     check_object_unsafe_self_trait_by_name(tcx, &trait_item);
     check_associated_item(tcx, trait_item.hir_id(), span, method_sig);
+
+    let encl_trait_hir_id = tcx.hir().get_parent_item(hir_id);
+    let encl_trait = tcx.hir().expect_item(encl_trait_hir_id);
+    if [tcx.lang_items().fn_trait(), tcx.lang_items().fn_mut_trait()]
+        .contains(&Some(encl_trait.def_id.to_def_id()))
+        && trait_item.ident.name.to_ident_string() == "call"
+    {
+        // We are looking at the `call` function of the `fn` or `fn_mut` lang item.
+        // Do some rudimentary sanity checking to avoid an ICE later (issue #83471).
+        if let Some(method_sig @ hir::FnSig { decl, .. }) = method_sig {
```

Comment:
Intrigued why you didn't do
```suggestion
        if let Some(hir::FnSig { decl, span .. }) = method_sig {
```